packages/polymath-ui/src/components/EthNetworkWrapper/actions.js

```diff
@@ -17,85 +32,119 @@ export type NetworkParams = {|
 // this has to be here for now
 const HARDCODED_NETWORK_ID = 15;
 
-export const CONNECTED = 'polymath-auth/CONNECTED';
+export const requestAuthorization = () => async (dispatch: Function) => {
+  try {
+    dispatch(fail(ERROR_ACCESS_REQUESTED));
+
+    // Request accounts access
+    await window.ethereum.enable();
+  } catch (e) {
+    // User denied access
+    dispatch(fail(ERROR_ACCESS_DENIED));
+  }
+};
+
 const connected = (params: NetworkParams) => ({ type: CONNECTED, params });
 
-export const FAIL = 'polymath-auth/FAIL';
-const fail = (error: number) => ({ type: FAIL, error });
+const fail = (error: number) => ({ type: FAILED, error });
 
-export type Action =
-  | ExtractReturn<typeof connected>
-  | ExtractReturn<typeof fail>;
+export const init = (networks: Array<string>) => async (dispatch: Function) => {
+  let web3;
+  let web3WS; // since MetaMask doesn't support WebSockets we need this extra client for events subscribing
+  let networkId = undefined;
 
-const web3 = new Web3();
-const web3WS = new Web3(); // since MetaMask doesn't support WebSockets we need this extra client for events subscribing
+  // Instantiate Web3 HTTP
+  if (window.ethereum) {
```

Comment:
This seems like a good opportunity for self-documenting code.

```suggestion
export const init = (networks: Array<string>) => async (dispatch: Function) => {
  let web3;
  let web3WS;
  let networkId;
  const newProviderInjected = !!window.ethereum;
  const oldProviderInjected = !!window.web3;
  
  if (newProviderInjected) {
    web3 = new Web3(window.ethereum);
  } else if (oldProviderInjected) {
    web3 = new Web3(window.web3.currentProvider);
  } else {
     return dispatch(fail(ERROR_NOT_INSTALLED)); 
  }
}
```